### PR TITLE
Integrate JsReporters

### PR DESCRIPTION
**Pull request by flore77**

### CODE REVIEW

**Comment by flore77** [comment]
@jzaefferer please review.

I spent a lot of time on debugging, I don't know what is happening in the MochaReporter with the `globalSuite.runtime`, I tested it separately and it works. 

Also please run the tests and tell me if everything is passing, yesterday when I wanted to test everything I received an error from Browserstack :laughing: and now when I am running them, my jobs are queued :disappointed:

**Comment by flore77** [comment]
:tada: My jobs have begun to run and all tests are passing.

**Comment by jzaefferer** [comment]
Can you provide some details on the `globalSuite.runtime` failure? What's failing specifically? What have you debugged already?

What is 'lib/_patch/my-mocha.js'? Doesn't look like its used anywhere.

Otherwise this is looking promising. There's still a lot of redundancy between each "plugin". We have to abstract that for this whole thing to make sense.

Can we make a single patch file, with something like this?

``` js
var runner;
if (window.QUnit) {
  runner = new JsReporters.QUnitAdapter(QUnit);
} else if (window.mocha) {
  runner = new JsReporters.MochaAdapter(mocha);
} else if (...
```

Then we can include a single file for qunit, mocha and jasmine2, and only load a separate one for jasmine1.

**Comment by flore77** [comment]
> What is 'lib/_patch/my-mocha.js'?

Yeah, it's a file I shouldn't have to commit, I will remove it. Sorry for that.

> What have you debugged already? 

I have debugged a lot why the new reporter was not functioning. Whit the runtime prop it got stuck and timeout. I put the reporter and js-reporters into the test file and run it in the browser, complete separately from browserstack, and it's functioning well, it computes the runtime.

> Can we make a single patch file, with something like this?

Sure, but this was the second step (from the issue description)  unifing all of them into a single file, I will do it.

**Comment by flore77** [comment]
@jzaefferer also a problem which I think it could be is that we are overriding Mocha's default reporter. I have seen that browserstack provides a video, as also screenshots of the testing. I think I have a fix, that can be applied in our adapter. Should I proceed into this ? (on the js-reporters repo) not here.

**Comment by jzaefferer** [comment]
Leaving the original mocha reporter intact makes sense. That would also be consistent with the other two frameworks, right?

**Comment by flore77** [comment]
Yes!

**Comment by flore77** [comment]
@jzaefferer check it out. The tests are passing, but the problem with the `runtime` for Mocha is still persisting :disappointed:

**Comment by jzaefferer** [comment]
I need some details on the problem with `runtime` and Mocha. Otherwise this looks pretty good.

**Comment by flore77** [comment]
Yes, I have discovered that the problem is not with the adapter or something like this, it is with `Internet Explorer 10`. I ran the tests in another browser (Mozilla 44) and it is functioning well. 

I checked and `Object.defineProperty` should be available in version 10. I cannot do more debug right now, I am working from a linux. I am trying to get a VM with Windows 7 and IE 10.

**Comment by jzaefferer** [comment]
Have you tried BrowserStack for that? The Live feature should work pretty well for that usecase.

**Comment by flore77** [comment]
Yes, but I don't how to load a local file into the browser :confused:

**Comment by jzaefferer** [comment]
They have a feature for local testing. Something tunnel? I'm sure it's documented somewhere. Used to work with a Chrome app, not sure if that's still required.

**Comment by flore77** [comment]
Yeah, I have manged it with BrowserStack, thanks!

image

That is the error, it is generated by the use of `for (let ... of ...)`. I have found out that this is happening also in other versions of IE and we should use a polyfill. I found a suggestion for [babel-polyfill](https://babeljs.io/docs/usage/polyfill/).

What should we do?

**Comment by flore77** [comment]
This looks good https://github.com/zloirock/core-js#commonjs, it is suggested on the above babel page, we can require only `es6.array.of`. Or it would be better to include all, to have no similar issues in the future?

**Comment by flore77** [comment]
Some interesting links:
- https://github.com/rollup/rollup/issues/616
- https://github.com/rollup/rollup/issues/766

**Comment by flore77** [comment]
Fixed also for `Mocha v2.4.5`. We still need to fix the thing with IE and then is good to land into master and pr to browserstack-runner repo.

**Comment by jzaefferer** [comment]
Let's start with the minimal polyfill.

**Comment by flore77** [comment]
Done. I think now we can do the pr against the browserstack repo :smiley:

**Comment by jzaefferer** [comment]
Great, go ahead!

File: lib/_patch/reporter.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    test.errors.forEach(function(error) {
+      tracebacks.push(error)
+    });
+  });
+
+  runner.on('runEnd', function(globalSuite) {
+    var results = {};
+
+    // TODO Investigate why is the runtime not functioning for Mocha.
```
Comment (jzaefferer) [inline lib/_patch/reporter.js]: Drop this, doesn't apply anymore

**Pull request merged**